CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,283 @@
+/// Take:
+//    - a PF candidate collection (which uses bugged HCAL respcorrs)
+//    - respCorrs values from fixed GT and bugged GT
+//  Produce:
+//    - a new PFCandidate collection containing the recalibrated PFCandidates in HF and where the neutral had pointing to problematic cells are removed
+//    - a second PFCandidate collection with just those discarded hadrons
+//    - a ValueMap<reco::PFCandidateRef> that maps the old to the new, and vice-versa
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+
+#include "FWCore/Framework/interface/Event.h"
+
+#include "CalibFormats/HcalObjects/interface/HcalDbService.h"
+#include "CalibFormats/HcalObjects/interface/HcalDbRecord.h"
+#include "CalibFormats/HcalObjects/interface/HcalCalibrations.h"
+#include "Geometry/CaloTopology/interface/HcalTopology.h"
+#include "Geometry/HcalTowerAlgo/interface/HcalGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloSubdetectorGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidateFwd.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+
+#include "DataFormats/Common/interface/View.h"
+#include "DataFormats/Common/interface/Association.h"
+#include <iostream>
+
+class PFCandidateRecalibrator : public edm::EDProducer {
+    public:
+        PFCandidateRecalibrator(const edm::ParameterSet&);
+        ~PFCandidateRecalibrator() override {};
+
+    private:
+        void beginRun(const edm::Run& iRun, edm::EventSetup const& iSetup) override;
+        void produce(edm::Event&, const edm::EventSetup&) override;
+
+        edm::EDGetTokenT<std::vector<reco::PFCandidate> > pfcandidates_;
+
+        std::vector<std::tuple<float,float,float>> badChHE_;
+        std::vector<std::tuple<int,int,int,float>> badChHF_;
+};
+
+PFCandidateRecalibrator::PFCandidateRecalibrator(const edm::ParameterSet &iConfig) :
+    pfcandidates_(consumes<std::vector<reco::PFCandidate>>(iConfig.getParameter<edm::InputTag>("pfcandidates")))
+{
+    produces<std::vector<reco::PFCandidate>>();
+    produces<std::vector<reco::PFCandidate>>("discarded");
+    produces<edm::ValueMap<reco::PFCandidateRef>>();
+}
+
+void PFCandidateRecalibrator::beginRun(const edm::Run &iRun, const edm::EventSetup &iSetup)
+{
+    //Get Calib Constants from current GT
```

Comment:
I don't see a reset to badCh vectors.
A job processing multiple runs will have these vectors grow at best redundantly, but at worst it will introduce dependence on the order of processed events if two runs have different sets of bad channels

Reply:
this is done in the last commit